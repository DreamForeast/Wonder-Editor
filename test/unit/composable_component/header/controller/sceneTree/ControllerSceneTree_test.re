open Wonder_jest;

open Expect;

open Expect.Operators;

open Sinon;

let _ =
  describe("controller sceneTree", () => {
    let sandbox = getSandboxDefaultVal();
    beforeEach(() => {
      sandbox := createSandbox();
      MainEditorSceneTool.initStateAndGl(~sandbox, ());
      MainEditorSceneTool.createDefaultScene(
        sandbox,
        MainEditorSceneTool.setFirstBoxTobeCurrentSceneTreeNode,
      );

      ControllerTool.stubRequestAnimationFrame(
        createEmptyStubWithJsObjSandbox(sandbox),
      );
      ControllerTool.run();
    });
    afterEach(() => restoreSandbox(refJsObjToSandbox(sandbox^)));

    describe("test set parent in engine", () => {
      test("no drag", () =>
        BuildComponentTool.buildSceneTree(
          TestTool.buildAppStateSceneGraphFromEngine(),
        )
        |> ReactTestTool.createSnapshotAndMatch
      );
      test(
        "drag treeNode into target treeNode, set draged gameObject's parent to be target gameObject",
        () => {
          let targetRunGameObject =
            StateLogicService.getRunEngineState()
            |> MainEditorSceneTool.getBoxByIndex(0);
          let dragedRunGameObject =
            StateLogicService.getRunEngineState()
            |> MainEditorSceneTool.getBoxByIndex(1);
          let targetEditGameObject =
            StateLogicService.getEditEngineState()
            |> MainEditorSceneTool.getBoxByIndex(0);
          let dragedEditGameObject =
            StateLogicService.getEditEngineState()
            |> MainEditorSceneTool.getBoxByIndex(1);
          let firstCubeDomIndex =
            SceneTreeNodeDomTool.OperateTwoLayer.getFirstCubeDomIndex();
          let secondCubeDomIndex =
            SceneTreeNodeDomTool.OperateTwoLayer.getSecondCubeDomIndex();

          let component =
            BuildComponentTool.buildSceneTree(
              TestTool.buildAppStateSceneGraphFromEngine(),
            );

          BaseEventTool.triggerComponentEvent(
            component,
            SceneTreeEventTool.triggerDragStart(secondCubeDomIndex),
          );
          BaseEventTool.triggerComponentEvent(
            component,
            SceneTreeEventTool.triggerDragEnter(firstCubeDomIndex),
          );
          BaseEventTool.triggerComponentEvent(
            component,
            SceneTreeEventTool.triggerDragDrop(firstCubeDomIndex),
          );
          (
            GameObjectUtils.getParent(
              dragedRunGameObject,
              StateLogicService.getRunEngineState(),
            ),
            GameObjectUtils.getParent(
              dragedEditGameObject,
              StateLogicService.getEditEngineState(),
            ),
          )
          |> expect == (targetRunGameObject, targetEditGameObject);
        },
      );
    });
  });