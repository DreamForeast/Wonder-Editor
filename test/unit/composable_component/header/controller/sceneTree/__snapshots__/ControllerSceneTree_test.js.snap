// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`controller sceneTree test set parent in engine no drag 1`] = `
<article
  className="wonder-sceneTree-component"
>
  <article
    className="wonder-tree"
  >
    <ul
      className="wonder-tree-node"
    >
      <li>
        <div
          className="item-triangle"
          onMouseDown={[Function]}
        >
          <img
            src="./public/img/down.png"
          />
        </div>
        <div
          className="draggable-container"
          draggable={true}
          onDragEnd={[Function]}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onMouseDown={[Function]}
          style={Object {}}
        >
          Scene
        </div>
      </li>
      <ul
        className="wonder-tree-node"
      >
        <li>
          <div
            className="item-triangle"
          />
          <div
            className="draggable-container"
            draggable={true}
            onDragEnd={[Function]}
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onMouseDown={[Function]}
            style={Object {}}
          >
            Camera
          </div>
        </li>
      </ul>
      <ul
        className="wonder-tree-node"
      >
        <li>
          <div
            className="item-triangle"
          />
          <div
            className="draggable-container select-active"
            draggable={true}
            onDragEnd={[Function]}
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onMouseDown={[Function]}
            style={Object {}}
          >
            Cube
          </div>
        </li>
      </ul>
      <ul
        className="wonder-tree-node"
      >
        <li>
          <div
            className="item-triangle"
          />
          <div
            className="draggable-container"
            draggable={true}
            onDragEnd={[Function]}
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onMouseDown={[Function]}
            style={Object {}}
          >
            Cube
          </div>
        </li>
      </ul>
      <ul
        className="wonder-tree-node"
      >
        <li>
          <div
            className="item-triangle"
          />
          <div
            className="draggable-container"
            draggable={true}
            onDragEnd={[Function]}
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onMouseDown={[Function]}
            style={Object {}}
          >
            Direction Light
          </div>
        </li>
      </ul>
    </ul>
  </article>
</article>
`;
